Query an IndexedView from a Primitive

donboie opened the pull request:
I'm not sure about the testing on this one

* C++ test bound via PrimiveBinding.cpp
* Create bindings for IndexedView & the new function on the Primitive class.

I'm a little unhappy about the unique_ptr return & installing an exception handler on each call, but can't think of anything better. 

This API is motivated by being able to use *Primitive::variableIndexedView* as a direct replacement for *Primitive::variableData*.

johnhaddon commented on include/IECoreScene/Primitive.inl:
```diff
@@ -38,6 +38,75 @@
+		}
+	}
+
+	if (!it->second.interpolation == requiredInterpolation)
```
Should this be `it->second.interpolation != requiredInterpolation`?

donboie replied:
Gah! There was a leading **!**  :man_facepalming:

johnhaddon commented on include/IECoreScene/Primitive.inl:
```diff
@@ -38,6 +38,75 @@
+	}
+	else
+	{
+		try
```
Rather than using exceptions for control flow, how about we just use the second IndexedView constructor instead?

```
// I'm taking some liberties with the return type - I'll discuss that in detail elsewhere.

typedef TypedData<vector<T>> DataType;
const DataType *data = runTimeCast<const DataType>( it->second.data.get() );
if( data )
{
     return IndexedView( data->readable(), it->second.indices ? it->second.indices->readable() : nullptr );
}
else if( throwIfInvalid )
{
    thow ...
}
else
{
    return boost::none;
}
```

donboie replied:
Perhaps it would be easier to use the TypedData as the template parameter as I won't have to jump through hoops to work out the TypedData from the value type.

johnhaddon replied:
Yeah, although I think looking at the call sites as you suggested below is more important - better to make it convenient for client code...

johnhaddon commented on include/IECoreScene/Primitive.h:
```diff
@@ -63,6 +65,24 @@ class IECORESCENE_API Primitive : public VisibleRenderable
+		/// If the required interpolation & datatype do not match then a nullptr unique_ptr is returned unless throwIfInvalid is true
+		/// The returned IndexedView lifetime must be bound by the primitive variable data it is viewing.
+		template<typename T>
+		std::unique_ptr<PrimitiveVariable::IndexedView<T> > variableIndexedView(
```
IndexedView is read-only, so I don't think we have any use for this non-const overload - the const one below should be sufficient.

johnhaddon replied:
This comment didn't seem to get attached to my review for some reason, so I'm just commenting to make sure you see it...

donboie commented:
> > This API is motivated by being able to use Primitive::variableIndexedView as a direct replacement for Primitive::variableData.

> With that in mind, should it be templated on the data type (e.g. FloatVectorData) rather than the the contained type (float)? I'm not sure either way...

Perhaps not direct replacement but a convienient one. I'll check the call sites I want to replace to see which parameter seems natural here.

donboie commented:
Addressed all review comments and squashed into 9f6acb8 

I've elected to use  data type as the template parameter as it's easy to typedef the required value_type for the IndexedView & perhaps *auto* might be convenient to refer to the optional IndexedView wihtout the hassle of specifying the type?

johnhaddon commented on include/IECoreScene/Primitive.h:
```diff
@@ -63,6 +67,17 @@ class IECORESCENE_API Primitive : public VisibleRenderable
 
+		/// To obtain an IndexedView onto a particular primitive variable.
+		/// An IndexedView allows access to both indexed & non indexed primitive variables using a common API.
+		/// If the required interpolation & datatype do not match then a nullptr unique_ptr is returned unless throwIfInvalid is true
```
Comment refers to previous `unique_ptr` approach.

johnhaddon commented on include/IECoreScene/Primitive.inl:
```diff
@@ -38,6 +38,67 @@
+		}
+	}
+
+	if( it->second.interpolation != requiredInterpolation )
```
To match `variableData()`, an unspecified interpolation should mean that any interpolation is accepted :
```
	if( requiredInterpolation!=PrimitiveVariable::Invalid && it->second.interpolation!=requiredInterpolation )
```

Sorry, I should have caught this at the same time as the original logic error I pointed out...
